### PR TITLE
libmysqlclient: Add v2 support

### PR DESCRIPTION
**Pull request by RubenRBS**
Specify library name and version:  **libmysqlclient/all**

Special thanks to @jcar87 for helping out in this one, it was a tough one to debug.
Along with https://github.com/conan-io/conan-center-index/pull/15247, this should get this library in pretty nice shape

Besides the general port changes, some extra ones are:

- Removes 404 mirrors, they do not exists at all
- Adds patches metadata
- Makes the `cmake` tool requirement optional on not having a new enough version already present. (Im not sure if downgrading the required version to the min one that fixes the original issue is ok. I'm happy to revert that change back to 3.24
- Removes the `package_id` method. This was only used to delete deprecated options, which I think is something that can be done in this case, but I might be missing something
- Thanks to @jcar87, patches the upstream `CMakeLists.txt` to actually look for `LZ4` paths. This might be worth to have as a patch instead of a finde and replace
- Remove 3.0.17. It was failing for some unknown reasons and didn't look like it was worth the effort to debug for now. Will maybe come back to this and re-add if necessary

### CODE REVIEW

File: recipes/libmysqlclient/all/conanfile.py [hunk 1]
```diff
@@ -100,160 +75,168 @@ def loose_lt_semver(v1, v2):
...
         rename(self, os.path.join(self.package_folder, "LICENSE"), os.path.join(self.package_folder, "licenses", "LICENSE"))
-        os.remove(os.path.join(self.package_folder, "README"))
+        rm(self, "README", self.package_folder)
         rm(self, "*.pdb", self.package_folder, recursive=True)
         rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
         rmdir(self, os.path.join(self.package_folder, "docs"))
         rmdir(self, os.path.join(self.package_folder, "share"))
         if self.settings.os == "Windows" and self.options.shared:
-            self.copy("*.dll", "bin", keep_path=False)
+            copy("*.dll", src=self.package_folder, dst="bin", keep_path=False)
```
Comment (jcar87) [inline recipes/libmysqlclient/all/conanfile.py:239]: mmm I think here the first argument needs to be self, the `src` is likely `self.build_folder` and the `dst` is `os.path.join(self.package_folder, "bin").

in conan 1.x self.copy is contextual which caused confusion :D

**Review by jwillikers — DISMISSED** [review]

**Review by StellaSmith — DISMISSED** [review]

File: recipes/libmysqlclient/all/test_v1_package/test_package.c [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#include <mysql.h>
```
Comment (SpaceIm) [inline recipes/libmysqlclient/all/test_v1_package/test_package.c:1]: code duplication can be avoided. You can have a look to recipe templates in conan-center doc.

File: recipes/libmysqlclient/all/conanfile.py [hunk 3]
```diff
@@ -100,160 +75,156 @@ def loose_lt_semver(v1, v2):
...
+        cmake.cache_variables["WITHOUT_SERVER"] = True
+        cmake.cache_variables["WITH_UNIT_TESTS"] = False
+        cmake.cache_variables["ENABLED_PROFILING"] = False
+        cmake.cache_variables["MYSQL_MAINTAINER_MODE"] = False
+        cmake.cache_variables["WIX_DIR"] = False
+
+        cmake.cache_variables["WITH_LZ4"] = "system"
+
+        cmake.cache_variables["WITH_ZSTD"] = "system"
+        cmake.cache_variables["ZSTD_INCLUDE_DIR"] = self.dependencies["zstd"].cpp_info.includedirs[0]
```
Comment (SpaceIm) [inline recipes/libmysqlclient/all/conanfile.py:200]:
```suggestion
        cmake.cache_variables["ZSTD_INCLUDE_DIR"] = self.dependencies["zstd"].cpp_info.aggregated_components().includedirs[0]
```

File: recipes/libmysqlclient/all/conanfile.py [hunk 4]
```diff
@@ -100,160 +75,156 @@ def loose_lt_semver(v1, v2):
...
-            self.build_requires("cmake/3.24.2")
+            self.tool_requires("cmake/3.24.2")
         if self.settings.os == "FreeBSD":
-            self.build_requires("pkgconf/1.9.3")
+            self.tool_requires("pkgconf/1.9.3")
 
     def source(self):
-        get(self, **self.conan_data["sources"][self.version],
-                  strip_root=True, destination=self._source_subfolder)
+        get(self, **self.conan_data["sources"][self.version], destination=self.source_folder, strip_root=True)
```
Comment (SpaceIm) [inline recipes/libmysqlclient/all/conanfile.py:109]:
```suggestion
        get(self, **self.conan_data["sources"][self.version], strip_root=True)
```

File: recipes/libmysqlclient/all/conanfile.py [hunk 5]
```diff
@@ -100,160 +75,156 @@ def loose_lt_semver(v1, v2):
...
-            cmake.definitions["ZSTD_INCLUDE_DIR"] = self.deps_cpp_info["zstd"].include_paths[0]
+            replace_in_file(self, os.path.join(self.source_folder, "libmysql", "CMakeLists.txt"),
+                            f"COMMAND {'libmysql_api_test'}",
+                            f"COMMAND DYLD_LIBRARY_PATH={os.path.join(self.build_folder, 'library_output_directory')} {os.path.join(self.build_folder, 'runtime_output_directory', 'libmysql_api_test')}")
+        replace_in_file(self, os.path.join(self.source_folder, "cmake", "install_macros.cmake"),
+                        "  INSTALL_DEBUG_SYMBOLS(",
+                        "  # INSTALL_DEBUG_SYMBOLS(")
+
+    def generate(self):
+        cmake = CMakeToolchain(self)
```
Comment (SpaceIm) [inline recipes/libmysqlclient/all/conanfile.py:186]: VirtualBuildEnv should be added since there are build requirements.

**Review by SpaceIm — CHANGES_REQUESTED** [review]

File: recipes/libmysqlclient/all/conanfile.py [hunk 6]
```diff
@@ -23,68 +22,44 @@ class LibMysqlClientCConan(ConanFile):
...
+            "Visual Studio": "16",
             "gcc": "7" if Version(self.version) >= "8.0.27" else "5.3",
             "clang": "6",
         }
 
     def export_sources(self):
-        self.copy("CMakeLists.txt")
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            self.copy(patch["patch_file"])
+        copy(self, "CMakeLists.txt", src=self.recipe_folder, dst=self.export_sources_folder)
```
[the remark this answers is not in the document]
Comment (RubenRBS) [reply recipes/libmysqlclient/all/conanfile.py:47]: Ooops, nice catch!

File: recipes/libmysqlclient/all/conanfile.py (hunk 3 above)
Comment (RubenRBS) [reply recipes/libmysqlclient/all/conanfile.py:200]: Wow TIL, thanks for the heads-up :)

File: recipes/libmysqlclient/all/conanfile.py (hunk 5 above)
Comment (RubenRBS) [reply recipes/libmysqlclient/all/conanfile.py:189]: Done, I'm still learning the ropes so thanks for checking I'm not missing things like these, I really appreciate it :)

**Comment by SpaceIm** [comment]
Please consider https://github.com/RubenRBS/conan-center-index/pull/3

**Comment by RubenRBS** [comment]
Thank you so much for the work to improve upon my changes :)

**Comment by SpaceIm** [comment]
I'm wondering: I remember that libmysqlclient has some try_run in its CMakeLists, this is why we prevent cross-build, but it may mean that VirtualRunEnv should be added to build scope to be robust to an "all shared" build.

File: recipes/libmysqlclient/all/conanfile.py [hunk 7]
```diff
@@ -24,67 +24,43 @@ class LibMysqlClientCConan(ConanFile):
...
-        "with_zlib": [True, False, "deprecated"],
     }
     default_options = {
         "shared": False,
         "fPIC": True,
-        "with_ssl": "deprecated",
-        "with_zlib": "deprecated",
     }
-
+    generators = "VirtualRunEnv", "VirtualBuildEnv", "CMakeDeps"
```
Comment (SpaceIm) [inline recipes/libmysqlclient/all/conanfile.py:32]: Not like this, you want fine-grained control of order and scope where these scripts will be called

**Review by danimtb — APPROVED** [review]

**Review by danimtb — APPROVED** [review]

File: recipes/libmysqlclient/all/conandata.yml [hunk 8]
```diff
@@ -17,28 +13,23 @@ sources:
...
-      base_path: "source_subfolder"
-  "8.0.17":
-    - patch_file: "patches/0001-find-cmake.patch"
-      base_path: "source_subfolder"
-    - patch_file: "patches/0002-dont-install-static-libraries+fix-mysql-config.patch"
-      base_path: "source_subfolder"
-    - patch_file: "patches/0003-msvc-install-no-pdb.patch"
-      base_path: "source_subfolder"
+      patch_description: "Fix macOS 12.0 version detection"
+      patch_type: "bugfix"
```
Comment (prince-chrismc) [inline recipes/libmysqlclient/all/conandata.yml:35]: Blocker. Incorrect usage of the patch fields, R the FM 🦒 lol

File: recipes/libmysqlclient/all/conanfile.py [hunk 9]
```diff
@@ -100,160 +77,169 @@ def loose_lt_semver(v1, v2):
...
+        vbenv = VirtualBuildEnv(self)
+        vbenv.generate()
+
+        if not cross_building(self):
+            vrenv = VirtualRunEnv(self)
+            vrenv.generate(scope="build")
+
+        tc = CMakeToolchain(self)
+        # Not used anywhere in the CMakeLists
+        tc.cache_variables["DISABLE_SHARED"] = not self.options.shared
```
Comment (prince-chrismc) [inline recipes/libmysqlclient/all/conanfile.py:197]: this is being patched in https://github.com/conan-io/conan-center-index/blob/22c9c9e77ebc2edfaae3888b9044c7987b5ff278/recipes/libmysqlclient/all/patches/0002-dont-install-static-libraries%2Bfix-mysql-config.patch#L9

**Review by prince-chrismc — APPROVED** [review]

**Pull request merged**